### ISSUE TITLE
Stop using `.gitignore` as a fallback for `.npmignore`

### ISSUE THREAD

**Issue by zkat** [npm/npm#12331]
I believe this feature, even though it's existed for a long time, has the potential to really mess with people.

I want to argue that the intentions of `.gitignore` and `.npmignore` are _fundamentally_ orthogonal to each other, and that while I do think this is _already_ a problem, I think it'll be an even bigger problem if/when we bring along specific changes and work more on how files/ignores are handles in npm. See https://github.com/npm/npm/pull/11669#issuecomment-209115892 specifically, which points out an actual problem that would happen.

In less fluffy terms: `.gitignore` is for excluding files from source control. npm, in its nature, very often deals with files that are excluded from source control (such as build artifacts), and is actually uninterested in having a lot of things that are _usually_ in source control (development-related configurations, build tooling and utilities, etc). This means that frankly, in most cases, I'd expect thorough `.npmignore` and `.gitignore` files to have _completely different contents_. I think this deprecation/change/feature is a good candidate for our next major release (and is trivial to actually implement), but has to be `breaking` because I'm sure someone out there will be fairly surprised by the change.

**Comment by iarna** [comment]
Removing this would literally break every repository I control.

**Comment by zkat** [comment]
Yeah. I suspect the scope is major enough that this is just something we can never manage. It's nice to actually recognize the scale of changing this particular thing, since at least I was talking about the idea of changing how `files` and `.npmignore` interact with each other: but making even that seemingly benign change means we have to look at `.gitignore`, and if we don't change that, we'll break a bunch of other people.

I'm sure there's a German word or something for the feeling I'm having about how the imo confusing and inconvenient current interaction between those two features simply has to be, because fixing even that small thing is intractable. Maybe there's some different way to think about it or maybe your particular experience is in the minority, but idk how we can even find out.

**Comment by ChALkeR** [comment]
@zkat The only way to make this change not leak potentially sensitive user data would be to refuse packaging and publishing if there is no `.npmignore` and `files` is not used.

**Comment by iarna** [comment]
TBH, if we were making radical changes I would prefer to ditch `.npmignore` entirely and just have `.gitignore` and `files`.

**Comment by ChALkeR** [comment]
@iarna To make that safe it should throw and refuse packaging/publishing if `.npmignore` is present, for quite a long time =).

**Comment by ljharb** [comment]
I really really don't want just `files` - whitelists make it far far too easy to accidentally forget to whitelist a new file, which causes the published package to break in a way that's really hard to detect from the publishing machine.

What I really want is `.gitignore` to always apply, but an `npm` config file that allows me to un-gitignore things, and extra-npm-ignore things - so that it's like a diff on the gitignore.

In other words, I definitely always want "all files included by default" because that's what should be published to `npm`, modulo specific things I want to exclude.

**Comment by mk-pmb** [comment]
@ iarna wrote:

> For my use cases (as a module author) I want to publish exactly the same things I check into git. The fact that npm doesn't quite do this was super frustrating to me early on (and continues to be a bugaboo for me).

Usually I'm in the same group, and want my package to contain exactly what's in my git repo. Now I'm writing a package where I care about the exact behavior, and learned about the magic just a few minutes ago while researching ways to save me from lots of `.npmignore -> .gitignore` symlinks.

My approach would be different, and if I'm not missing something, less breaking:

  * [ ] Add a `package.json` field `excludeLists` which is an array of filenames whose lines will be interpreted, in each subdirectory where they exist, as **additional** entries to `.npmignore`.
    * (Name based on `excludes` or `excludeFiles`, which I thought to be a black-list twin of `files` but cannot find it.)
    * [ ] https://docs.npmjs.com/files/package.json#files should have a link to explain this, ideally near the search keyword "exclude" or containing it.
    * [ ] If `excludeLists` is set to any list (even an empty one) or `false`, it means per-package opt-in to the new behavior and implies opt-out of `.gitignore` magic.
    * I can set it to `[ ".gitignore" ]` and expect to then be able to rely on never needing any `.npmignore`, even in a future npm with possibly changed magic defaults, except:
    * [ ] If there's an `.npmignore`, its entries will "win" over the additional files.
    * Someone might want to invent special `.npmignore` syntax to revoke entire categories of lines in the additional files without having to write an explicit list of override entries.
    * [ ] Since there may be a change of default behavior in the future, and bugs might arise from the change, add an `.npmrc` option that says which behavior the npm user expects. If npm encounters an expected behavior setting that's not supported in its current version, fail with a helpful message.

To help against accidential publishing of secret files, I suggest:

  * [ ] Specify an `.npmrc` option of how fatal a warning/error it shall be when I try to publish a package where npm's effective list of ignored files differs from git's effective list of ignored files.
    * [ ] If not yet in the client, please add it.
    * [ ] If set to "I don't care", don't waste resources on trying to compare the lists.
    * [ ] https://docs.npmjs.com/files/package.json#files should have a link to explain this, ideally near the search keyword "exclude" or containing it.

**Comment by kyleoliveiro** [comment]
To not break things, maybe make this an opt-in-only feature via a setting in `package.json`?

**Comment by binki** [comment]
For many people, the `.gitignore` fallback works and is desired behavior. For people using built artifacts, they can probably just as easily build `.npmignore` too. So I think this behavior should just be kept as-is.

Example of how to (non-portably) build `.npmignore` by deriving from `.gitignore`: https://github.com/binki/promise-delay-ts/blob/1b1fc0c311e3442ff8050b95e66c1d1a8850ee98/package.json#L10
